Enhance admin media UI with dominant color support

pbearne opened the pull request:
Added inline CSS and JavaScript to display dominant colors and transparency in the WordPress admin area for attachment previews. This enhancement allows the media library to utilize dominant color and transparency data, providing visual feedback on image properties. Updated version to 1.1.2 to reflect these changes.

## Summary

<!-- Please reference the issue this PR fixes. If this PR does not fix the entire issue, change this to Addresses #... instead. -->
Fixes #

## Relevant technical choices

<!-- Please describe your changes. -->



<!--
For maintainers only, please make sure:

- PR has a `[Type]` label.
- PR has a plugin-specific milestone, or the `no milestone` label if it does not apply to any specific plugin.
- PR has a changelog-friendly title, or the `skip changelog` label if it should not be mentioned in the plugin's changelog.
-->

Commit:
Enhance admin UI with dominant color support.

Added inline CSS and JavaScript to display dominant colors and transparency in the WordPress admin area for attachment previews. This enhancement allows the media library to utilize dominant color and transparency data, providing visual feedback on image properties. Updated version to 1.1.2 to reflect these changes.

pbearne committed:
Merge branch 'trunk' into 354-media-model

Commit:
Bump version to 1.1.3 for dominant color images plugin

Update version numbers and associated comments from 1.1.2 to 1.1.3 for consistency across the plugin files. This change ensures that the codebase reflects the correct version for recent updates or bug fixes included in this release.

pbearne committed:
Update hooks.php

Commit:
Fix code indentation and spacing issues

Corrected inconsistent indentation and spacing to improve code readability and maintainability in hooks.php. These changes ensure a cleaner code style without altering the functionality or logic.

Commit:
Fix code indentation and spacing issues

Corrected inconsistent indentation and spacing to improve code readability and maintainability in hooks.php. These changes ensure a cleaner code style without altering the functionality or logic.

Commit:
Add return types to dominant color functions

This commit specifies return types for several functions in the dominant-color-images plugin. The `dominant_color_admin_inline_style` and `dominant_color_admin_script` functions now explicitly return void, and the `dominant_color_prepare_attachment_for_js` function now returns an array. Adding these return types improves code clarity and aligns with current PHP best practices.

Commit:
Remove unnecessary return void annotations

The @return void annotations were removed from the PHPDoc comments in two functions within the hooks.php file. These annotations were redundant, as the functions' return types are already declared with "void" in the function signatures. This cleanup enhances the readability and maintains consistency across the codebase.

Commit:
Refactor type hints in function docblock.

Updated the function docblock type hints for arrays to use generics, enhancing code clarity and consistency. This change does not affect functionality but improves the readability and maintenance of the codebase.

Commit:
Refactor type hints in function docblock.

Updated the function docblock type hints for arrays to use generics, enhancing code clarity and consistency. This change does not affect functionality but improves the readability and maintenance of the codebase.

joemcgill commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,76 @@ function dominant_color_render_generator(): void {
+ * to apply background color based on the dominant color for attachment previews
+ * in the WordPress admin interface.
+ *
+ * @since 1.1.3
```
Just in case, these could be set to n.e.x.t. and auto-updated to the correct version when this is packaged in a release:

```suggestion
 * @since n.e.x.t
```

🔢 Same throughout

joemcgill commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,76 @@ function dominant_color_render_generator(): void {
+ * @return array<mixed> The modified response array with added dominant color and transparency information.
+ */
+function dominant_color_prepare_attachment_for_js( array $response, WP_Post $attachment, array $meta ): array {
+	unset( $attachment );
```
Why is unsetting this necessary? Seems like this could be omitted.

joemcgill commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,76 @@ function dominant_color_render_generator(): void {
+function dominant_color_prepare_attachment_for_js( array $response, WP_Post $attachment, array $meta ): array {
+	unset( $attachment );
+	$response['dominantColor'] = '';
+	if ( isset( $meta['dominant_color'] ) ) {
```
It may be worth adding some more validation rather than just ensuring these database values are set. Additionally, let's sanitize these values before returning them in the response if possible.

pbearne replied:
done this

joemcgill commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,76 @@ function dominant_color_render_generator(): void {
+	$response['hasTransparencyClass'] = '';
+	if ( isset( $meta['has_transparency'] ) ) {
+		$response['hasTransparency']      = $meta['has_transparency'];
+		$response['hasTransparencyClass'] = $meta['has_transparency'] ? 'has-transparency' : 'not-transparent';
```
This could be inferred on the front end and not returned as part of the response.

pbearne replied:
I belive we need this

joemcgill replied:
It's fine to create the class on the server and pass it as part of the response, but just to clarify what I meant, in the script that gets output in `dominant_color_admin_script()`, you could do something like this (untested):

```js
	<script>
		(function() {
			var s = jQuery( '#tmpl-attachment' )
			var transparencyClassName = 'true' === data.hasTransparency ? 'has-transparency' : 'not-transparent';
			var n = s[0].innerText.replace( '{{ data.orientation }}"', '{{ data.orientation }} {{ transparencyClassName }}"	data-dominant-color="{{ data.dominantColor }}" data-has-transparency="{{ data.hasTransparency }}" style="--dominant-color: #{{ data.dominantColor }};"')
			s.replaceWith( '<script type="text/html" id="tmpl-attachment">' + n  );
		}());
	</script>
```

Also, it looks like only the `has-transparency` class is ever referenced by the custom CSS that gets added here, so if could be simplified to:

```
var transparencyClassName = 'true' === data.hasTransparency ? 'has-transparency' : '';
```

or if this remains in PHP:

```php
$response['hasTransparencyClass'] = $meta['has_transparency'] ? 'has-transparency' : '';
```

pbearne replied:
made the change to use JS for the class

joemcgill commented on plugins/dominant-color-images/readme.txt:
```diff
@@ -47,6 +47,12 @@ Contributions are always welcome! Learn more about how to get involved in the [C
 
 == Changelog ==
 
+= 1.1.3 =
```
I'm not positive, but I think these will get auto-generated during the deployment process, so no need to include them in the PR. Thanks for being so detailed about this though! ✨

westonruter replied:
That's right. Running `npm run readme` will automatically add the changelog for PRs in the milestone.

joemcgill reviewed (changes requested):
Thanks @pbearne. This looks mostly good to go for me. Left a few suggestions

Commit:
Ensure correct color and attribute sanitization.

Updated the dominant color and transparency properties in hooks.php to use proper sanitization functions. This change helps prevent potential issues with unsanitized data affecting the media library display. Also, adjusted version references for an upcoming release.

westonruter commented on plugins/dominant-color-images/load.php:
```diff
@@ -5,7 +5,7 @@
  * Requires at least: 6.6
  * Requires PHP: 7.2
- * Version: 1.1.2
+ * Version: 1.1.3
```
I suggest perhaps reverting the version changes since we typically bump all those when doing a release. For example, at the time of release we can evaluate what type of version bump is warranted, whether a patch or a minor version bump.

```suggestion
 * Version: 1.1.2
```

westonruter commented on plugins/dominant-color-images/load.php:
```diff
@@ -25,7 +25,7 @@
 }
 
-define( 'DOMINANT_COLOR_IMAGES_VERSION', '1.1.2' );
+define( 'DOMINANT_COLOR_IMAGES_VERSION', '1.1.3' );
```
```suggestion
define( 'DOMINANT_COLOR_IMAGES_VERSION', '1.1.2' );
```

westonruter commented on plugins/dominant-color-images/readme.txt:
```diff
@@ -2,7 +2,7 @@
 Contributors: wordpressdotorg
 Tested up to: 6.7
-Stable tag:   1.1.2
+Stable tag:   1.1.3
```
```suggestion
Stable tag:   1.1.2
```

westonruter commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,76 @@ function dominant_color_render_generator(): void {
+	$response['hasTransparency']      = '';
+	$response['hasTransparencyClass'] = '';
+	if ( isset( $meta['has_transparency'] ) ) {
+		$response['hasTransparency']      = esc_attr( $meta['has_transparency'] );
```
Is `esc_attr()` correct here? As I recall the data here will be passed back as JSON via admin-ajax, and won't this run the risk of accidentally encoding HTML entities? If the the value should be a boolean, would this make more sense as:

```suggestion
		$response['hasTransparency']      = (bool) $meta['has_transparency'];
```

Otherwise, if if it is supposed to be a string then literals could be used as in the line below.

westonruter commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,76 @@ function dominant_color_render_generator(): void {
+ *
+ * @since n.e.x.t
+ *
+ * @param array<mixed> $response   The current response array for the attachment.
+ * @param WP_Post      $attachment The attachment post object.
+ * @param array<mixed> $meta       The attachment metadata.
+ * @return array<mixed> The modified response array with added dominant color and transparency information.
+ */
+function dominant_color_prepare_attachment_for_js( array $response, WP_Post $attachment, array $meta ): array {
```
Other plugins could be doing bad things and not returning an `array` as they should:

```suggestion
 * @param array<mixed>|mixed $response   The current response array for the attachment.
 * @param WP_Post            $attachment The attachment post object.
 * @param array<mixed>       $meta       The attachment metadata.
 * @return array<mixed> The modified response array with added dominant color and transparency information.
 */
function dominant_color_prepare_attachment_for_js( $response, WP_Post $attachment, array $meta ): array {
	if ( ! is_array( $response ) ) {
		$response = array();
	}
```

pbearne committed:
Update plugins/dominant-color-images/load.php

Co-authored-by: Weston Ruter <westonruter@google.com>

pbearne committed:
Update plugins/dominant-color-images/load.php

Co-authored-by: Weston Ruter <westonruter@google.com>

pbearne committed:
Update plugins/dominant-color-images/readme.txt

Co-authored-by: Weston Ruter <westonruter@google.com>

pbearne committed:
Update plugins/dominant-color-images/hooks.php

Co-authored-by: Weston Ruter <westonruter@google.com>

pbearne committed:
Update plugins/dominant-color-images/hooks.php

Co-authored-by: Weston Ruter <westonruter@google.com>

Commit:
Refactor transparency class assignment logic

Reorganize the logic for assigning the transparency class name by moving it from the server-side to the client-side script. This change improves code clarity by handling conditional class name assignments directly within the client-side context, thereby simplifying the server-side response handling.

Commit:
Fix inconsistent spacing in hooks.php file

Remove unnecessary whitespace around the assignment operator in the 'hasTransparency' key modification. This change ensures code consistency and readability in the dominant-color-images plugin.

pbearne committed:
Merge branch 'trunk' into 354-media-model

westonruter commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,79 @@ function dominant_color_render_generator(): void {
+	<script>
+		(function() {
+			var s = jQuery( '#tmpl-attachment' )
+			var transparencyClassName = 'true' === data.hasTransparency ? 'has-transparency' : 'not-transparent';
```
Something doesn't seem right here. The `data` variable is being used in the script itself and not in the template. Won't it be `undefined`?

westonruter replied:
Per below:

```suggestion
```

westonruter commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,79 @@ function dominant_color_render_generator(): void {
+		(function() {
+			var s = jQuery( '#tmpl-attachment' )
+			var transparencyClassName = 'true' === data.hasTransparency ? 'has-transparency' : 'not-transparent';
+			var n = s[0].innerText.replace( '{{ data.orientation }}"', '{{ data.orientation }} {{ transparencyClassName }}"	data-dominant-color="{{ data.dominantColor }}" data-has-transparency="{{ data.hasTransparency }}" style="--dominant-color: #{{ data.dominantColor }};"')
```
Isn't this what is intended instead?

```suggestion
			var n = s[0].innerText.replace( '{{ data.orientation }}"', '{{ data.orientation }} {{ data.hasTransparency ? 'has-transparency' : 'not-transparent' }}"	data-dominant-color="{{ data.dominantColor }}" data-has-transparency="{{ data.hasTransparency }}" style="--dominant-color: #{{ data.dominantColor }};"')
```

westonruter replied:
Also, just updated the suggestion to replace `'true' === data.hasTransparency` with just `data.hasTransparency`, as as far as I can tell, this is a `bool` and not a `string` since it is being casted below:

```php
$response['hasTransparency'] = (bool) $meta['has_transparency'];
```

westonruter commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,79 @@ function dominant_color_render_generator(): void {
+ */
+function dominant_color_admin_script(): void {
+	?>
+	<script>
+		(function() {
+			var s = jQuery( '#tmpl-attachment' )
+			var transparencyClassName = 'true' === data.hasTransparency ? 'has-transparency' : 'not-transparent';
+			var n = s[0].innerText.replace( '{{ data.orientation }}"', '{{ data.orientation }} {{ transparencyClassName }}"	data-dominant-color="{{ data.dominantColor }}" data-has-transparency="{{ data.hasTransparency }}" style="--dominant-color: #{{ data.dominantColor }};"')
+			s.replaceWith( '<script type="text/html" id="tmpl-attachment">' + n  );
+		}());
+	</script>
+	<?php
+}
+add_action( 'admin_print_footer_scripts', 'dominant_color_admin_script', 1000 );
```
Incorporating my above feedback, I think this can be further simplified:

1. Use a module so that we don't worry about needing the IIFE to eliminate global scope pollution.
2. Use a module so that we're guaranteed that the script only executes once the DOM has fully loaded, so no need for specifying a 1000 priority.
3. Eliminate the use of jQuery in favor of using plain JS calls.
4. Use `textContent` instead of `innerText` since it seems to be the better choice here.
5. Update the text content instead of replacing the entire element.

```suggestion
	<script type="module">
		const tmpl = document.getElementById( 'tmpl-attachment' );
		if ( tmpl ) {
			tmpl.textContent = tmpl.textContent.replace(
				'{{ data.orientation }}"',
				'{{ data.orientation }} {{ data.hasTransparency ? \'has-transparency\' : \'not-transparent\' }}"	data-dominant-color="{{ data.dominantColor }}" data-has-transparency="{{ data.hasTransparency }}" style="--dominant-color: #{{ data.dominantColor }};"'
			);
		}
	</script>
	<?php
}
add_action( 'admin_print_footer_scripts', 'dominant_color_admin_script' );
```

pbearne replied:
love the new JS

westonruter commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,79 @@ function dominant_color_render_generator(): void {
+	}
+
+	$response['dominantColor'] = '';
+	if ( isset( $meta['dominant_color'] ) && str_contains( '#', $meta['dominant_color'] ) ) {
```
```suggestion
	if ( isset( $meta['dominant_color'] ) && str_starts_with( '#', $meta['dominant_color'] ) ) {
```

pbearne committed:
Update plugins/dominant-color-images/hooks.php

Co-authored-by: Weston Ruter <westonruter@google.com>

pbearne committed:
Update plugins/dominant-color-images/hooks.php

Co-authored-by: Weston Ruter <westonruter@google.com>

pbearne committed:
Update hooks.php

cleaner JS code

westonruter commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,80 @@ function dominant_color_render_generator(): void {
+	</script>
+	<?php
+}
+add_action( 'admin_print_footer_scripts', 'dominant_color_admin_script', 1000 );
```
There shouldn't be a need for the late priority now that a module script is being used.

```suggestion
add_action( 'admin_print_footer_scripts', 'dominant_color_admin_script' );
```

westonruter commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,80 @@ function dominant_color_render_generator(): void {
+	<script type="module">
+		const tmpl = document.getElementById( 'tmpl-attachment' );
+		if ( tmpl ) {
+			tmpl.textContent = tmpl.textContent.replace(
+				'{{ data.orientation }}"',
+				'{{ data.orientation }} {{ data.hasTransparency ? \'has-transparency\' : \'not-transparent\' }}"	data-dominant-color="{{ data.dominantColor }}" data-has-transparency="{{ data.hasTransparency }}" style="--dominant-color: #{{ data.dominantColor }};"'
+			);
```
As I understand, this will be injecting additional attributes at the end of this `DIV`: https://github.com/WordPress/wordpress-develop/blob/0d428194d1e6d6f1285d45cafd84b18f514070c1/src/wp-includes/media-template.php#L596

This seems like it could be a bit brittle, like what if someone has that text string show up somewhere else? Or what if another plugin already appended the `class` with some other text than `{{ data.orientation }}`? Would it be better to instead just look for the first `DIV` and then append the attributes? So something like this:

```suggestion
		tmpl.textContent = tmpl.textContent.replace( /^\s*<div[^>]*?(?=>)/, ( match ) => {
			let replaced = match.replace( /\sclass="/, " class=\"{{ data.hasTransparency ? \'has-transparency\' : \'not-transparent\' }} " );
			replaced += ' data-dominant-color="{{ data.dominantColor }}"';
			replaced += ' data-has-transparency="{{ data.hasTransparency }}"';
			replaced += ' style="--dominant-color: #{{ data.dominantColor }};"'; // TODO: Potentially there could be a style attribute as well!
			return replaced;
		} );
```

Granted, this is probably not going to happen.

Too bad we don't have the HTML Tag Processor in JavaScript!

pbearne committed:
Update plugins/dominant-color-images/hooks.php

Co-authored-by: Weston Ruter <westonruter@google.com>

pbearne committed:
Update plugins/dominant-color-images/hooks.php

Co-authored-by: Weston Ruter <westonruter@google.com>

pbearne committed:
Merge branch 'trunk' into 354-media-model

westonruter commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,83 @@ function dominant_color_render_generator(): void {
+	<script type="module">
+		const tmpl = document.getElementById( 'tmpl-attachment' );
+		if ( tmpl ) {
+		tmpl.textContent = tmpl.textContent.replace( /^\s*<div[^>]*?(?=>)/, ( match ) => {
+			let replaced = match.replace( /\sclass="/, " class=\"{{ data.hasTransparency ? \'has-transparency\' : \'not-transparent\' }} " );
+			replaced += ' data-dominant-color="{{ data.dominantColor }}"';
+			replaced += ' data-has-transparency="{{ data.hasTransparency }}"';
+			replaced += ' style="--dominant-color: #{{ data.dominantColor }};"'; // TODO: Potentially there could be a style attribute as well!
+			return replaced;
+		} );
```
Fixing indentation:

```suggestion
			tmpl.textContent = tmpl.textContent.replace( /^\s*<div[^>]*?(?=>)/, ( match ) => {
				let replaced = match.replace( /\sclass="/, " class=\"{{ data.hasTransparency ? \'has-transparency\' : \'not-transparent\' }} " );
				replaced += ' data-dominant-color="{{ data.dominantColor }}"';
				replaced += ' data-has-transparency="{{ data.hasTransparency }}"';
				replaced += ' style="--dominant-color: #{{ data.dominantColor }};"'; // TODO: Potentially there could be a style attribute as well!
				return replaced;
			} );
```

mukeshpanchal27 commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,83 @@ function dominant_color_render_generator(): void {
+		$response['dominantColor'] = sanitize_hex_color( $meta['dominant_color'] );
+	}
+	$response['hasTransparency']      = '';
+	$response['hasTransparencyClass'] = '';
```
Do we need these?

pbearne committed:
Update plugins/dominant-color-images/hooks.php

Co-authored-by: Weston Ruter <westonruter@google.com>

pbearne committed:
Update hooks.php

westonruter commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,82 @@ function dominant_color_render_generator(): void {
+	if ( isset( $meta['dominant_color'] ) && str_starts_with( '#', $meta['dominant_color'] ) ) {
+		$response['dominantColor'] = sanitize_hex_color( $meta['dominant_color'] );
+	}
+	$response['hasTransparency']      = '';
```
This should fix the PHPCS complaint:

```suggestion
	$response['hasTransparency'] = '';
```

pbearne committed:
Update plugins/dominant-color-images/hooks.php

Co-authored-by: Weston Ruter <westonruter@google.com>

pbearne committed:
Merge branch 'trunk' into 354-media-model

westonruter commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,82 @@ function dominant_color_render_generator(): void {
+ */
+function dominant_color_admin_inline_style(): void {
+	$handle = 'dominant-color-admin-styles';
+	// PHPCS ignore reason: Version not used since this handle is only registered for adding an inline style.
+	// phpcs:ignore WordPress.WP.EnqueuedResourceParameters.MissingVersion
```
Nit: The rationale can be added to the same comment per https://github.com/squizlabs/PHP_CodeSniffer/wiki/Advanced-Usage#:~:text=If%20required%2C%20you%20can%20add%20a%20note%20explaining%20why%20the%20file%20is%20being%20ignored%20by%20using%20the%20%2D%2D%20separator.

```suggestion
	// phpcs:ignore WordPress.WP.EnqueuedResourceParameters.MissingVersion -- Version not used since this handle is only registered for adding an inline style.
```

westonruter commented on plugins/dominant-color-images/hooks.php:
```diff
@@ -186,3 +186,82 @@ function dominant_color_render_generator(): void {
+				let replaced = match.replace( /\sclass="/, " class=\"{{ data.hasTransparency ? \'has-transparency\' : \'not-transparent\' }} " );
+				replaced += ' data-dominant-color="{{ data.dominantColor }}"';
+				replaced += ' data-has-transparency="{{ data.hasTransparency }}"';
+				replaced += ' style="--dominant-color: #{{ data.dominantColor }};"'; // TODO: Potentially there could be a style attribute as well!
```
This is untested, but it should resolve the remaining todo (and again, this may be overkill):

```suggestion
				let hasStyleAttr = false;
				const colorStyle = '--dominant-color: #{{ data.dominantColor }};';
				replaced = replaced.replace( /\sstyle="/, ( styleMatch ) => {
					hasStyleAttr = true;
					return styleMatch + colorStyle;
				} );
				if ( ! hasStyleAttr ) {
					replaced += ` style="${colorStyle}"`;
				}
```

pbearne committed:
Update plugins/dominant-color-images/hooks.php

Co-authored-by: Weston Ruter <westonruter@google.com>

westonruter committed:
Merge branch 'trunk' into 354-media-model

westonruter committed:
Combine comments

westonruter commented:
I'm seeing the dominant color on the frontend:

```html
<img data-dominant-color="070606" data-has-transparency="false" style="--dominant-color: #070606;" fetchpriority="high" decoding="async" width="1024" height="403" src="http://localhost:8888/wp-content/uploads/2024/12/Screenshot-2024-12-09-12.43.02-1024x403.png" alt="" class="wp-image-402 not-transparent" srcset="http://localhost:8888/wp-content/uploads/2024/12/Screenshot-2024-12-09-12.43.02-1024x403.png 1024w, http://localhost:8888/wp-content/uploads/2024/12/Screenshot-2024-12-09-12.43.02-300x118.png 300w, http://localhost:8888/wp-content/uploads/2024/12/Screenshot-2024-12-09-12.43.02-768x302.png 768w, http://localhost:8888/wp-content/uploads/2024/12/Screenshot-2024-12-09-12.43.02.png 1294w" sizes="(max-width: 1024px) 100vw, 1024px">
```

But on the Media screen and in the Media modal, I'm not seeing anything added:

```html
<div class="not-transparent attachment-preview js--select-attachment type-image subtype-png landscape" data-dominant-color="" data-has-transparency="false" style="--dominant-color: #;">
```

Note the dominant color is missing from the underlying data. In looking at the Ajax response, I see:

```json
{
    "id": 402,
    "title": "Screenshot 2024-12-09 12.43.02",
    "filename": "Screenshot-2024-12-09-12.43.02.png",
    "url": "http:\/\/localhost:8888\/wp-content\/uploads\/2024\/12\/Screenshot-2024-12-09-12.43.02.png",
    "link": "http:\/\/localhost:8888\/screenshot-2024-12-09-12-43-02\/",
    "alt": "",
    "author": "1",
    "description": "",
    "caption": "",
    "name": "screenshot-2024-12-09-12-43-02",
    "status": "inherit",
    "uploadedTo": 0,
    "date": 1734116876000,
    "modified": 1734116876000,
    "menuOrder": 0,
    "mime": "image\/png",
    "type": "image",
    "subtype": "png",
    "icon": "http:\/\/localhost:8888\/wp-includes\/images\/media\/default.svg",
    "dateFormatted": "December 13, 2024",
    "nonces": {
        "update": "6ec61bcfb7",
        "delete": "92edbf45a7",
        "edit": "dccdfde8bb"
    },
    "editLink": "http:\/\/localhost:8888\/wp-admin\/post.php?post=402&action=edit",
    "meta": false,
    "authorName": "admin",
    "authorLink": "http:\/\/localhost:8888\/wp-admin\/profile.php",
    "filesizeInBytes": 92142,
    "filesizeHumanReadable": "90 KB",
    "context": "",
    "height": 509,
    "width": 1294,
    "orientation": "landscape",
    "sizes": {
        "thumbnail": {
            "height": 150,
            "width": 150,
            "url": "http:\/\/localhost:8888\/wp-content\/uploads\/2024\/12\/Screenshot-2024-12-09-12.43.02-150x150.png",
            "orientation": "landscape"
        },
        "medium": {
            "height": 118,
            "width": 300,
            "url": "http:\/\/localhost:8888\/wp-content\/uploads\/2024\/12\/Screenshot-2024-12-09-12.43.02-300x118.png",
            "orientation": "landscape"
        },
        "large": {
            "height": 238,
            "width": 604,
            "url": "http:\/\/localhost:8888\/wp-content\/uploads\/2024\/12\/Screenshot-2024-12-09-12.43.02-1024x403.png",
            "orientation": "landscape"
        },
        "full": {
            "url": "http:\/\/localhost:8888\/wp-content\/uploads\/2024\/12\/Screenshot-2024-12-09-12.43.02.png",
            "height": 509,
            "width": 1294,
            "orientation": "landscape"
        }
    },
    "compat": {
        "item": "",
        "meta": ""
    },
    "dominantColor": "",
    "hasTransparency": false
}
```

westonruter commented:
In looking at the underlying attachment meta, the `dominant_color` does not start with `#`"

```json
{
    "dominant_color": "070606",
    "has_transparency": false
}
```

westonruter committed:
Account for dominant_color not being prefixed by hash

westonruter committed:
Avoid printing --dominant-color style when there is no dominant color

westonruter committed:
Fix regex for dominant_color

westonruter commented:
Seems like this is now working!

Before:

[Screen recording 2024-12-13 11.52.07.webm](https://github.com/user-attachments/assets/cf4381c7-e69a-4dcf-9f31-a60c39d8dcf8)

![image](https://github.com/user-attachments/assets/e26b63a2-bd48-401a-91cb-aaee8044a1ab)

After:

[Screen recording 2024-12-13 11.53.06.webm](https://github.com/user-attachments/assets/98d6c280-59ec-4be2-9fd1-688b68e42903)

![image](https://github.com/user-attachments/assets/4b6b901b-25a1-4036-a8f4-2acb24d51248)

westonruter committed:
Add reference to related core function

joemcgill reviewed (approved):
Looking good to me. Thanks @pbearne!
